### PR TITLE
[DUOS-2473] Elasticsearch dataset query

### PR DESCRIPTION
**Pull request by fboulnois**
### Addresses

https://broadworkbench.atlassian.net/browse/DUOS-2473

### Summary

Code to query the elasticsearch endpoint.

----
Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

- Label PR with a Jira ticket number and include a link to the ticket
- Label PR with a security risk modifier [no, low, medium, high]
- PR describes scope of changes
- Get a minimum of one thumbs worth of review, preferably two if enough team members are available
- Get PO sign-off for all non-trivial UI or workflow changes
- Verify all tests go green
- Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/main/java/org/broadinstitute/consent/http/service/ElasticSearchService.java [hunk 1]
```diff
@@ -66,6 +68,37 @@ public Response indexDatasetTerms(List<DatasetTerm> datasets) throws IOException
     return esClient.performRequest(bulkRequest);
   }
 
+  public boolean validateQuery(String query) throws IOException {
```
Comment (rushtong) [inline src/main/java/org/broadinstitute/consent/http/service/ElasticSearchService.java:97]: This is nice, thank you!

**Review by kaitlynoflaherty — APPROVED** [review]
Looks good

File: src/main/java/org/broadinstitute/consent/http/resources/DatasetResource.java [hunk 2]
```diff
@@ -660,15 +659,9 @@ public Response searchDatasets(
   public Response searchDatasetIndex(@Auth AuthUser authUser, String query) {
```
Comment (rushtong) [inline src/main/java/org/broadinstitute/consent/http/resources/DatasetResource.java:667]: I think we need a `@PermitAll` line for this method - I definitely missed this in my earlier review.

File: src/main/java/org/broadinstitute/consent/http/service/ElasticSearchService.java [hunk 3]
```diff
@@ -66,6 +68,37 @@ public Response indexDatasetTerms(List<DatasetTerm> datasets) throws IOException
...
+    if (!validateQuery(query)) {
+      throw new IOException("Invalid Elasticsearch query");
+    }
+
+    Request searchRequest = new Request(
+        HttpMethod.GET,
+        "/" + esConfig.getDatasetIndexName() + "/_search");
+    searchRequest.setEntity(new NStringEntity(query, ContentType.APPLICATION_JSON));
+
+    return esClient.performRequest(searchRequest);
```
Comment (rushtong) [inline src/main/java/org/broadinstitute/consent/http/service/ElasticSearchService.java]: I'm seeing a `JerseyGsonProvider` exception parsing the response for what should be a valid query
```
{ "query": { "query_string": { "query": "(GRU) OR (HMB)" } } }
```
...
```
consent  | ERROR [2023-07-24 13:28:17,461] io.dropwizard.jersey.errors.LoggingExceptionMapper: Error handling a request: 9db3ef52a2d8eada
consent  | ! java.lang.reflect.InaccessibleObjectException: Unable to make field final byte[] java.nio.ByteBuffer.hb accessible: module java.base does not "opens java.nio" to unnamed module @3b4ce44c
consent  | ! at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(Unknown Source)
consent  | ! at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(Unknown Source)
consent  | ! at java.base/java.lang.reflect.Field.checkCanSetAccessible(Unknown Source)
consent  | ! at java.base/java.lang.reflect.Field.setAccessible(Unknown Source)
consent  | ! at com.google.gson.internal.reflect.ReflectionHelper.makeAccessible(ReflectionHelper.java:35)
consent  | ! ... 84 common frames omitted
```

**Review by rushtong — CHANGES_REQUESTED** [review]
Have not been able to get this to function yet - need to debug the parsing error some more. I think we could also use some example(s) and expanded description in the swagger docs to make it easier for UI devs to make use of this endpoint.

File: src/main/resources/assets/schemas/DatasetSearch.yaml [hunk 4]
```diff
@@ -0,0 +1,33 @@
+type: object
+properties:
```
Comment (rushtong) [inline src/main/resources/assets/schemas/DatasetSearch.yaml:2]: There should be a lot more potential properties here (see [DatasetSearchTerm](https://github.com/DataBiosphere/consent/blob/develop/src/main/java/org/broadinstitute/consent/http/models/DatasetSearchTerm.java)). I think we still have a bug where a lot of those fields are not actually being sent to ES right now.

**Review by rushtong — APPROVED** [review]
Looks and works great! We can likely use some additional documentation somewhere for UI dev. I was able to see valid results for a variety of queries, very exciting!
```
{ "query": { "match": { "dataUse.primary.code": "DS"} } }

{ "query": { "match": { "dataUse.secondary.code": "NCU"} } }

{ "query": { "match": { "approvedUserIds": 4949} } }

{
  "query": {
    "bool": {
      "must": [
        { "match": { "approvedUserIds": 4949} },
        { "match": { "dataUse.primary.code": "GRU"} }
      ]
    }
  }
}
```

**Comment by rushtong** [comment]
Approval caveat ... let's get the tests to pass.

**Review by rushtong — APPROVED** [review]
Thank you for the changes 👍🏽

**Pull request merged**